components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,366 @@
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Returns Result of UnicodeSet from a single pair of integers defining a range
+    ///
+    /// `start`: inclusive, `end`: exclusive
+    ///
+    /// Returns an error if the range is invalid (out of order and out of bounds).
+    ///
+    /// Example Call: `UnicodeSet::from_range(&0, &15)`
+    pub fn from_range(start: u32, end: u32) -> Result<UnicodeSet, Box<dyn Error>> {
+        if start > end {
+            return Err(format!("Range is out of order. start: {} end: {}", start, end).into());
+        }
+        if start < CODEPOINT_MIN || end > CODEPOINT_MAX {
+            return Err(format!(
+                "Range is out of bounds. start: {}, min: {}, end: {}, max: {}",
+                start, CODEPOINT_MIN, end, CODEPOINT_MAX
+            )
+            .into());
+        }
+        Ok(UnicodeSet {
+            inv_list: vec![start, end],
+        })
+    }
+
+    /// Returns UnicodeSet spanning entire Unicode range
+    ///
+    /// The range spans from `0x0 -> 0x10FFFF` inclusive
+    pub fn all() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, CODEPOINT_MAX + 1],
+        }
+    }
+
+    /// Returns UnicodeSet spanning BMP range
+    ///
+    /// The range spans from `0x0 -> 0xFFFF` inclusive
+    pub fn bmp() -> UnicodeSet {
+        UnicodeSet {
+            inv_list: vec![CODEPOINT_MIN, BMP_MAX + 1],
+        }
+    }
+
+    /// Yields an iterator of start and stop points of ranges in the UnicodeSet
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu4x_unicodeset::UnicodeSet;
+    /// let example = UnicodeSet::new("4 0 10 15 20");
+    /// let mut example_iter = example.iter();
+    /// example_iter.next(); // => 0
+    /// example_iter.next(); // => 10
+    /// example_iter.next(); // => 10, etc.
+    /// ```
+    pub fn iter(&self) -> Iter<u32> {
+        self.inv_list.iter()
+    }
+
+    /// Returns the number of elements of the UnicodeSet
+    ///
+    /// Returns an error if a complete range is not defined in the UnicodeSet
+    pub fn size(&self) -> usize {
+        if self.is_empty() {
+            return 0;
+        }
+        let end_point_sum: u32 = self.iter().skip(1).step_by(2).sum::<u32>();
+        let start_point_sum: u32 = self.iter().step_by(2).sum::<u32>();
+        (end_point_sum - start_point_sum) as usize
+    }
+
+    /// Returns whether or not the UnicodeSet is empty
+    pub fn is_empty(&self) -> bool {
+        self.inv_list.len() < 2 // unsure if this is appropriate definition of just self.inv_list.is_empty()
+    }
+
+    /// Wrapper for contains
+    ///
+    /// Takes in a single code point `query`, and a closure `condition`
+    /// to see if the `query` is located in the inversion list.
+    ///
+    /// Example:
+    ///
+    /// `let condition_closure = |_: usize| -> bool {true};`
+    /// `self.contains(10, condition_closure);`
+    fn contains<C>(&self, query: u32, condition: C) -> bool
```

Comment:
Here and elsewhere: Since this is a UnicodeSet, I would make these functions like `contains` accept a `char` instead of a `u32`.  You can cast to a `u32` once inside the function.

Reply:
agreed and changed.